{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "LoadError",
     "evalue": "UndefVarError: @nbinclude not defined",
     "output_type": "error",
     "traceback": [
      "UndefVarError: @nbinclude not defined",
      ""
     ]
    }
   ],
   "source": [
    "using NBInclude\n",
    "@nbinclude(\"network.ipynb\")\n",
    "@nbinclude(\"genome.ipynb\")\n",
    "@nbinclude(\"nnode.ipynb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutable struct NNode \n",
    "    activation_count::Int32\n",
    "    last_activation::Float32\n",
    "    last_activation2::Float32\n",
    "    \n",
    "    nodetrait = Trait[]\n",
    "    trait_id::Int32\n",
    "    dup = NNode[]\n",
    "    analogue = NNode[]\n",
    "    override::Bool\n",
    "    override_value::Float32\n",
    "    frozen::Bool\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class NNode {\n",
    "\n",
    "    int activation_count;  // keeps track of which activation the node is currently in\n",
    "    double last_activation; // Holds the previous step's activation for recurrency\n",
    "    double last_activation2; // Holds the activation BEFORE the prevous step's\n",
    "    // This is necessary for a special recurrent case when the innode\n",
    "    // of a recurrent link is one time step ahead of the outnode.\n",
    "    // The innode then needs to send from TWO time steps ago\n",
    "\n",
    "    Trait *nodetrait; // Points to a trait of parameters\n",
    "\n",
    "    int trait_id;  // identify the trait derived by this node\n",
    "\n",
    "    NNode *dup;       // Used for Genome duplication\n",
    "\n",
    "    NNode *analogue;  // Used for Gene decoding\n",
    "\n",
    "    bool override; // The NNode cannot compute its own output- something is overriding it\n",
    "\n",
    "    double override_value; // Contains the activation value that will override this node's activation\n",
    "\n",
    "    // Pointer to the Sensor corresponding to this Body.\n",
    "    //Sensor* mySensor;\n",
    "\n",
    "public:\n",
    "    bool frozen; // When frozen, cannot be mutated (meaning its trait pointer is fixed)\n",
    "\n",
    "    functype ftype; // type is either SIGMOID ..or others that can be added\n",
    "    nodetype type; // type is either NEURON or SENSOR \n",
    "\n",
    "    double activesum;  // The incoming activity before being processed \n",
    "    double activation; // The total activation entering the NNode \n",
    "    bool active_flag;  // To make sure outputs are active\n",
    "\n",
    "    // NOT USED IN NEAT - covered by \"activation\" above\n",
    "    double output;  // Output of the NNode- the value in the NNode \n",
    "\n",
    "    // ************ LEARNING PARAMETERS *********** \n",
    "    // The following parameters are for use in    \n",
    "    //   neurons that learn through habituation,\n",
    "    //   sensitization, or Hebbian-type processes  \n",
    "\n",
    "    double params[NEAT::num_trait_params];\n",
    "\n",
    "    std::vector<Link*> incoming; // A list of pointers to incoming weighted signals from other nodes\n",
    "    std::vector<Link*> outgoing;  // A list of pointers to links carrying this node's signal\n",
    "\n",
    "    // These members are used for graphing with GTK+/GDK\n",
    "    std::vector<double> rowlevels;  // Depths from output where this node appears\n",
    "    int row;  // Final row decided upon for drawing this NNode in\n",
    "    int ypos;\n",
    "    int xpos;\n",
    "\n",
    "    int node_id;  // A node can be given an identification number for saving in files\n",
    "\n",
    "    nodeplace gen_node_label;  // Used for genetic marking of nodes\n",
    "\n",
    "    NNode(nodetype ntype,int nodeid);\n",
    "\n",
    "    NNode(nodetype ntype,int nodeid, nodeplace placement);\n",
    "\n",
    "    // Construct a NNode off another NNode for genome purposes\n",
    "    NNode(NNode *n,Trait *t);\n",
    "\n",
    "    // Construct the node out of a file specification using given list of traits\n",
    "    NNode (const char *argline, std::vector<Trait*> &traits);\n",
    "\n",
    "    // Copy Constructor\n",
    "    NNode (const NNode& nnode);\n",
    "\n",
    "    ~NNode();\n",
    "\n",
    "    // Just return activation for step\n",
    "    double get_active_out();\n",
    "\n",
    "    // Return activation from PREVIOUS time step\n",
    "    double get_active_out_td();\n",
    "\n",
    "    // Returns the type of the node, NEURON or SENSOR\n",
    "    const nodetype get_type();\n",
    "\n",
    "    // Allows alteration between NEURON and SENSOR.  Returns its argument\n",
    "    nodetype set_type(nodetype);\n",
    "\n",
    "    // If the node is a SENSOR, returns true and loads the value\n",
    "    bool sensor_load(double);\n",
    "\n",
    "    // Adds a NONRECURRENT Link to a new NNode in the incoming List\n",
    "    void add_incoming(NNode*,double);\n",
    "\n",
    "    // Adds a Link to a new NNode in the incoming List\n",
    "    void add_incoming(NNode*,double,bool);\n",
    "\n",
    "    // Recursively deactivate backwards through the network\n",
    "    void flushback();\n",
    "\n",
    "    // Verify flushing for debugging\n",
    "    void flushback_check(std::vector<NNode*> &seenlist);\n",
    "\n",
    "    // Print the node to a file\n",
    "    void  print_to_file(std::ostream &outFile);\n",
    "void print_to_file(std::ofstream &outFile);\n",
    "\n",
    "    // Have NNode gain its properties from the trait\n",
    "    void derive_trait(Trait *curtrait);\n",
    "\n",
    "    // Returns the gene that created the node\n",
    "    NNode *get_analogue();\n",
    "\n",
    "    // Force an output value on the node\n",
    "    void override_output(double new_output);\n",
    "\n",
    "    // Tell whether node has been overridden\n",
    "    bool overridden();\n",
    "\n",
    "    // Set activation to the override value and turn off override\n",
    "    void activate_override();  \n",
    "\n",
    "    // Writes back changes weight values into the genome\n",
    "    // (Lamarckian trasnfer of characteristics)\n",
    "    void Lamarck();\n",
    "\n",
    "    //Find the greatest depth starting from this neuron at depth d\n",
    "    int depth(int d,Network *mynet); \n",
    "\n",
    "};"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function NNode(nodetype ntype,int nodeid) {\n",
    "    active_flag=false;\n",
    "    activesum=0;\n",
    "    activation=0;\n",
    "    output=0;\n",
    "    last_activation=0;\n",
    "    last_activation2=0;\n",
    "    type=ntype; //NEURON or SENSOR type\n",
    "    activation_count=0; //Inactive upon creation\n",
    "    node_id=nodeid;\n",
    "    ftype=SIGMOID;\n",
    "    nodetrait=0;\n",
    "    gen_node_label=HIDDEN;\n",
    "    dup=0;\n",
    "    analogue=0;\n",
    "    frozen=false;\n",
    "    trait_id=1;\n",
    "    override=false;\n",
    "}\n",
    "\n",
    "NNode::NNode(nodetype ntype,int nodeid, nodeplace placement) {\n",
    "    active_flag=false;\n",
    "    activesum=0;\n",
    "    activation=0;\n",
    "    output=0;\n",
    "    last_activation=0;\n",
    "    last_activation2=0;\n",
    "    type=ntype; //NEURON or SENSOR type\n",
    "    activation_count=0; //Inactive upon creation\n",
    "    node_id=nodeid;\n",
    "    ftype=SIGMOID;\n",
    "    nodetrait=0;\n",
    "    gen_node_label=placement;\n",
    "    dup=0;\n",
    "    analogue=0;\n",
    "    frozen=false;\n",
    "    trait_id=1;\n",
    "    override=false;\n",
    "}\n",
    "\n",
    "NNode::NNode(NNode *n,Trait *t) {\n",
    "    active_flag=false;\n",
    "    activation=0;\n",
    "    output=0;\n",
    "    last_activation=0;\n",
    "    last_activation2=0;\n",
    "    type=n->type; //NEURON or SENSOR type\n",
    "    activation_count=0; //Inactive upon creation\n",
    "    node_id=n->node_id;\n",
    "    ftype=SIGMOID;\n",
    "    nodetrait=0;\n",
    "    gen_node_label=n->gen_node_label;\n",
    "    dup=0;\n",
    "    analogue=0;\n",
    "    nodetrait=t;\n",
    "    frozen=false;\n",
    "    if (t!=0)\n",
    "        trait_id=t->trait_id;\n",
    "    else trait_id=1;\n",
    "    override=false;\n",
    "}\n",
    "\n",
    "NNode::NNode (const char *argline, std::vector<Trait*> &traits) {\n",
    "    int traitnum;\n",
    "    std::vector<Trait*>::iterator curtrait;\n",
    "\n",
    "    activesum=0;\n",
    "\n",
    "    std::stringstream ss(argline);\n",
    "    //char curword[128];\n",
    "    //char delimiters[] = \" \\n\";\n",
    "    //int curwordnum = 0;\n",
    "\n",
    "    //Get the node parameters\n",
    "    //strcpy(curword, NEAT::getUnit(argline, curwordnum++, delimiters));\n",
    "    //node_id = atoi(curword);\n",
    "    //strcpy(curword, NEAT::getUnit(argline, curwordnum++, delimiters));\n",
    "    //traitnum = atoi(curword);\n",
    "    //strcpy(curword, NEAT::getUnit(argline, curwordnum++, delimiters));\n",
    "    //type = (nodetype)atoi(curword);\n",
    "    //strcpy(curword, NEAT::getUnit(argline, curwordnum++, delimiters));\n",
    "    //gen_node_label = (nodeplace)atoi(curword);\n",
    "\n",
    "    int nodety, nodepl;\n",
    "    ss >> node_id >> traitnum >> nodety >> nodepl;\n",
    "    type = (nodetype)nodety;\n",
    "    gen_node_label = (nodeplace)nodepl;\n",
    "\n",
    "    // Get the Sensor Identifier and Parameter String\n",
    "    // mySensor = SensorRegistry::getSensor(id, param);\n",
    "    frozen=false;  //TODO: Maybe change\n",
    "\n",
    "    //Get a pointer to the trait this node points to\n",
    "    if (traitnum==0) nodetrait=0;\n",
    "    else {\n",
    "        curtrait=traits.begin();\n",
    "        while(((*curtrait)->trait_id)!=traitnum)\n",
    "            ++curtrait;\n",
    "        nodetrait=(*curtrait);\n",
    "        trait_id=nodetrait->trait_id;\n",
    "    }\n",
    "\n",
    "    override=false;\n",
    "}\n",
    "\n",
    "// This one might be incomplete\n",
    "NNode::NNode (const NNode& nnode){\n",
    "    active_flag = nnode.active_flag;\n",
    "    activesum = nnode.activesum;\n",
    "    activation = nnode.activation;\n",
    "    output = nnode.output;\n",
    "    last_activation = nnode.last_activation;\n",
    "    last_activation2 = nnode.last_activation2;\n",
    "    type = nnode.type; //NEURON or SENSOR type\n",
    "    activation_count = nnode.activation_count; //Inactive upon creation\n",
    "    node_id = nnode.node_id;\n",
    "    ftype = nnode.ftype;\n",
    "    nodetrait = nnode.nodetrait;\n",
    "    gen_node_label = nnode.gen_node_label;\n",
    "    dup = nnode.dup;\n",
    "    analogue = nnode.dup;\n",
    "    frozen = nnode.frozen;\n",
    "    trait_id = nnode.trait_id;\n",
    "    override = nnode.override;\n",
    "}\n",
    "\n",
    "NNode::~NNode() {\n",
    "    std::vector<Link*>::iterator curlink;\n",
    "\n",
    "    //Kill off all incoming links\n",
    "    for(curlink=incoming.begin();curlink!=incoming.end();++curlink) {\n",
    "        delete (*curlink);\n",
    "    }\n",
    "    //if (nodetrait!=0) delete nodetrait;\n",
    "}\n",
    "\n",
    "//Returns the type of the node, NEURON or SENSOR\n",
    "const nodetype NNode::get_type() {\n",
    "    return type;\n",
    "}\n",
    "\n",
    "//Allows alteration between NEURON and SENSOR.  Returns its argument\n",
    "nodetype NNode::set_type(nodetype newtype) {\n",
    "    type=newtype;\n",
    "    return newtype;\n",
    "}\n",
    "\n",
    "//If the node is a SENSOR, returns true and loads the value\n",
    "bool NNode::sensor_load(double value) {\n",
    "    if (type==SENSOR) {\n",
    "\n",
    "        //Time delay memory\n",
    "        last_activation2=last_activation;\n",
    "        last_activation=activation;\n",
    "\n",
    "        activation_count++;  //Puts sensor into next time-step\n",
    "        activation=value;\n",
    "        return true;\n",
    "    }\n",
    "    else return false;\n",
    "}\n",
    "\n",
    "// Note: NEAT keeps track of which links are recurrent and which\n",
    "// are not even though this is unnecessary for activation.\n",
    "// It is useful to do so for 2 other reasons: \n",
    "// 1. It makes networks visualization of recurrent networks possible\n",
    "// 2. It allows genetic control of the proportion of connections\n",
    "//    that may become recurrent\n",
    "\n",
    "// Add an incoming connection a node\n",
    "void NNode::add_incoming(NNode *feednode,double weight,bool recur) {\n",
    "    Link *newlink=new Link(weight,feednode,this,recur);\n",
    "    incoming.push_back(newlink);\n",
    "    (feednode->outgoing).push_back(newlink);\n",
    "}\n",
    "\n",
    "// Nonrecurrent version\n",
    "void NNode::add_incoming(NNode *feednode,double weight) {\n",
    "    Link *newlink=new Link(weight,feednode,this,false);\n",
    "    incoming.push_back(newlink);\n",
    "    (feednode->outgoing).push_back(newlink);\n",
    "}\n",
    "\n",
    "// Return activation currently in node, if it has been activated\n",
    "double NNode::get_active_out() {\n",
    "    if (activation_count>0)\n",
    "        return activation;\n",
    "    else return 0.0;\n",
    "}\n",
    "\n",
    "// Return activation currently in node from PREVIOUS (time-delayed) time step,\n",
    "// if there is one\n",
    "double NNode::get_active_out_td() {\n",
    "    if (activation_count>1)\n",
    "        return last_activation;\n",
    "    else return 0.0;\n",
    "}\n",
    "\n",
    "// This recursively flushes everything leading into and including this NNode, including recurrencies\n",
    "void NNode::flushback() {\n",
    "    std::vector<Link*>::iterator curlink;\n",
    "\n",
    "    //A sensor should not flush black\n",
    "    if (type!=SENSOR) {\n",
    "\n",
    "        if (activation_count>0) {\n",
    "            activation_count=0;\n",
    "            activation=0;\n",
    "            last_activation=0;\n",
    "            last_activation2=0;\n",
    "        }\n",
    "\n",
    "        //Flush back recursively\n",
    "        for(curlink=incoming.begin();curlink!=incoming.end();++curlink) {\n",
    "            //Flush the link itself (For future learning parameters possibility) \n",
    "            (*curlink)->added_weight=0;\n",
    "            if ((((*curlink)->in_node)->activation_count>0))\n",
    "                ((*curlink)->in_node)->flushback();\n",
    "        }\n",
    "    }\n",
    "    else {\n",
    "        //Flush the SENSOR\n",
    "        activation_count=0;\n",
    "        activation=0;\n",
    "        last_activation=0;\n",
    "        last_activation2=0;\n",
    "\n",
    "    }\n",
    "\n",
    "}\n",
    "\n",
    "// This recursively checks everything leading into and including this NNode, \n",
    "// including recurrencies\n",
    "// Useful for debugging\n",
    "void NNode::flushback_check(std::vector<NNode*> &seenlist) {\n",
    "    std::vector<Link*>::iterator curlink;\n",
    "    //int pause;\n",
    "    std::vector<Link*> innodes=incoming;\n",
    "    std::vector<NNode*>::iterator location;\n",
    "\n",
    "    if (!(type==SENSOR)) {\n",
    "\n",
    "\n",
    "        //std::cout<<\"ALERT: \"<<this<<\" has activation count \"<<activation_count<<std::endl;\n",
    "        //std::cout<<\"ALERT: \"<<this<<\" has activation  \"<<activation<<std::endl;\n",
    "        //std::cout<<\"ALERT: \"<<this<<\" has last_activation  \"<<last_activation<<std::endl;\n",
    "        //std::cout<<\"ALERT: \"<<this<<\" has last_activation2  \"<<last_activation2<<std::endl;\n",
    "\n",
    "        if (activation_count>0) {\n",
    "            std::cout<<\"ALERT: \"<<this<<\" has activation count \"<<activation_count<<std::endl;\n",
    "        }\n",
    "\n",
    "        if (activation>0) {\n",
    "            std::cout<<\"ALERT: \"<<this<<\" has activation  \"<<activation<<std::endl;\n",
    "        }\n",
    "\n",
    "        if (last_activation>0) {\n",
    "            std::cout<<\"ALERT: \"<<this<<\" has last_activation  \"<<last_activation<<std::endl;\n",
    "        }\n",
    "\n",
    "        if (last_activation2>0) {\n",
    "            std::cout<<\"ALERT: \"<<this<<\" has last_activation2  \"<<last_activation2<<std::endl;\n",
    "        }\n",
    "\n",
    "        for(curlink=innodes.begin();curlink!=innodes.end();++curlink) {\n",
    "            location = std::find(seenlist.begin(),seenlist.end(),((*curlink)->in_node));\n",
    "            if (location==seenlist.end()) {\n",
    "                seenlist.push_back((*curlink)->in_node);\n",
    "                ((*curlink)->in_node)->flushback_check(seenlist);\n",
    "            }\n",
    "        }\n",
    "\n",
    "    }\n",
    "    else {\n",
    "        //Flush_check the SENSOR\n",
    "\n",
    "\n",
    "        std::cout<<\"sALERT: \"<<this<<\" has activation count \"<<activation_count<<std::endl;\n",
    "        std::cout<<\"sALERT: \"<<this<<\" has activation  \"<<activation<<std::endl;\n",
    "        std::cout<<\"sALERT: \"<<this<<\" has last_activation  \"<<last_activation<<std::endl;\n",
    "        std::cout<<\"sALERT: \"<<this<<\" has last_activation2  \"<<last_activation2<<std::endl;\n",
    "\n",
    "\n",
    "        if (activation_count>0) {\n",
    "            std::cout<<\"ALERT: \"<<this<<\" has activation count \"<<activation_count<<std::endl;\n",
    "        }\n",
    "\n",
    "        if (activation>0) {\n",
    "            std::cout<<\"ALERT: \"<<this<<\" has activation  \"<<activation<<std::endl;\n",
    "        }\n",
    "\n",
    "        if (last_activation>0) {\n",
    "            std::cout<<\"ALERT: \"<<this<<\" has last_activation  \"<<last_activation<<std::endl;\n",
    "        }\n",
    "\n",
    "        if (last_activation2>0) {\n",
    "            std::cout<<\"ALERT: \"<<this<<\" has last_activation2  \"<<last_activation2<<std::endl;\n",
    "        }\n",
    "\n",
    "    }\n",
    "\n",
    "}\n",
    "\n",
    "// Reserved for future system expansion\n",
    "void NNode::derive_trait(Trait *curtrait) {\n",
    "\n",
    "    if (curtrait!=0) {\n",
    "        for (int count=0;count<NEAT::num_trait_params;count++)\n",
    "            params[count]=(curtrait->params)[count];\n",
    "    }\n",
    "    else {\n",
    "        for (int count=0;count<NEAT::num_trait_params;count++)\n",
    "            params[count]=0;\n",
    "    }\n",
    "\n",
    "    if (curtrait!=0)\n",
    "        trait_id=curtrait->trait_id;\n",
    "    else trait_id=1;\n",
    "\n",
    "}\n",
    "\n",
    "// Returns the gene that created the node\n",
    "NNode *NNode::get_analogue() {\n",
    "    return analogue;\n",
    "}\n",
    "\n",
    "// Force an output value on the node\n",
    "void NNode::override_output(double new_output) {\n",
    "\toverride_value=new_output;\n",
    "    override=true;\n",
    "}\n",
    "\n",
    "// Tell whether node has been overridden\n",
    "bool NNode::overridden() {\n",
    "    return override;\n",
    "}\n",
    "\n",
    "// Set activation to the override value and turn off override\n",
    "void NNode::activate_override() {\n",
    "    activation=override_value;\n",
    "    override=false;\n",
    "}\n",
    "\n",
    "\n",
    "void NNode::print_to_file(std::ofstream &outFile) {\n",
    "  outFile<<\"node \"<<node_id<<\" \";\n",
    "  if (nodetrait!=0) outFile<<nodetrait->trait_id<<\" \";\n",
    "  else outFile<<\"0 \";\n",
    "  outFile<<type<<\" \";\n",
    "  outFile<<gen_node_label<<std::endl;\n",
    "}\n",
    "\n",
    "\n",
    "void NNode::print_to_file(std::ostream &outFile) {\n",
    "    //outFile<<\"node \"<<node_id<<\" \";\n",
    "    //if (nodetrait!=0) outFile<<nodetrait->trait_id<<\" \";\n",
    "    //else outFile<<\"0 \";\n",
    "    //outFile<<type<<\" \";\n",
    "    //outFile<<gen_node_label<<std::endl;\n",
    "\n",
    "    char tempbuf[128];\n",
    "    sprintf(tempbuf, \"node %d \", node_id);\n",
    "    outFile << tempbuf;\n",
    "\n",
    "    if (nodetrait != 0) {\n",
    "        char tempbuf2[128];\n",
    "        sprintf(tempbuf2, \"%d \", nodetrait->trait_id);\n",
    "        outFile << tempbuf2;\n",
    "    }\n",
    "    else outFile << \"0 \";\n",
    "\n",
    "    char tempbuf2[128];\n",
    "    sprintf(tempbuf2, \"%d %d\\n\", type, gen_node_label);\n",
    "    outFile << tempbuf2;\n",
    "}\n",
    "\n",
    "//Find the greatest depth starting from this neuron at depth d\n",
    "int NNode::depth(int d, Network *mynet) {\n",
    "  std::vector<Link*> innodes=this->incoming;\n",
    "  std::vector<Link*>::iterator curlink;\n",
    "  int cur_depth; //The depth of the current node\n",
    "  int max=d; //The max depth\n",
    "\n",
    "  if (d>100) {\n",
    "    //std::cout<<mynet->genotype<<std::endl;\n",
    "    //std::cout<<\"** DEPTH NOT DETERMINED FOR NETWORK WITH LOOP\"<<std::endl;\n",
    "    return 10;\n",
    "  }\n",
    "\n",
    "  //Base Case\n",
    "  if ((this->type)==SENSOR)\n",
    "    return d;\n",
    "  //Recursion\n",
    "  else {\n",
    "\n",
    "    for(curlink=innodes.begin();curlink!=innodes.end();++curlink) {\n",
    "      cur_depth=((*curlink)->in_node)->depth(d+1,mynet);\n",
    "      if (cur_depth>max) max=cur_depth;\n",
    "    }\n",
    "  \n",
    "    return max;\n",
    "\n",
    "  } //end else\n",
    "\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.0.0",
   "language": "julia",
   "name": "julia-1.0"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.0.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
